Windows: make Sys.remove and Unix.unlink delete symlinks to directories

MisterDA opened the pull request:
On Windows, calling `_wunlink` on a path may fail with `EACCES` if [1]:
1. the path specifies a read-only file;
2. the path specifies a directory.

On Windows, a symbolic link to a directory is actually a directory
with an associated reparse point. The consequence is that `_wunlink`
will fail on them. However, POSIX `unlink(2)` specifies [2] that it
will gladly unlink symlinks to directories. POSIX `unlink(2)` also
forbids deletion of directories.

To emulate that behaviour, we wrap the `_wunlink` call. If it fails
with `EACCES`, we have to check whether the path is a directory with
an associated reparse point, and that the type of the reparse point is
a symbolic link. If so, we call `_wrmdir` on the path.

[1]: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/unlink-wunlink?view=msvc-160
[2]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html

We cannot define `unlink_os` to `caml_win32_unlink` as ocamlyacc
includes `caml/misc.h`. It would need the definition from `win32.c`,
which isn't linked in ocamlyacc.

Introduce the `caml/winsupport.h` header to hold common code between
the runtime, lib-sys, and win32unix. This is also mandated by the name
conflict between `ATOM` from `<windef.h>` and `caml/instruct.h`.

The PR used to change the errno from `EACCES` to `EPERM` when trying
to unlink a directory (on Windows) but that was reverted during the
review.

MisterDA committed:
Windows: Sys.remove, Unix.unlink now remove symlinks to directories

On Windows, calling `_wunlink` on a path may fail with `EACCES` if [1]:
1. the path specifies a read-only file;
2. the path specifies a directory.

On Windows, a symbolic link to a directory is actually a directory
with an associated reparse point. The consequence is that `_wunlink`
will fail on them. However, POSIX `unlink(2)` specifies [2] that it
will gladly unlink symlinks to directories. POSIX `unlink(2)` also
forbids deletion of directories.

To emulate that behaviour, we wrap the `_wunlink` call. If it fails
with `EACCES`, we have to check whether the path is a directory with
an associated reparse point, and that the type of the reparse point is
a symbolic link. If so, we call `_wrmdir` on the path.

[1]: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/unlink-wunlink?view=msvc-160
[2]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html

We cannot define `unlink_os` to `caml_win32_unlink` as ocamlyacc
includes `caml/misc.h`. It would need the definition from `win32.c`,
which isn't linked in ocamlyacc.

Introduce the `caml/winsupport.h` header to hold common code between
the runtime, lib-sys, and win32unix. This is also mandated by the name
conflict between `ATOM` from `<windef.h>` and `caml/instruct.h`.

MisterDA committed:
Test Sys.remove, Unix.unlink behavior on symlinks and directories

POSIX mandates that unlink(2) is be able to delete symlinks, including
symlinks to directories, but doesn't delete directories
themselves (that's the purpose of rmdir).

MisterDA commented:
THB "POSIX compliance" is a bit strongly-worded, I'm mostly interested by the bug-fix of removing symlinks to directories. Seeing that Windows's `_unlink` returning `EACCES` on directories is documented in Windows and OCaml, and that even Linux returns `EISDIR` instead of following POSIX's `EPERM`, it may be too much here. I can revert that part of the PR.

dra27 commented on otherlibs/win32unix/unix.ml:
```diff
@@ -361,6 +361,7 @@ external isatty : file_descr -> bool = "unix_isatty"
 
 (* Operations on file names *)
 
+(* Unlink is redefined after rmdir definition. *)
```
I'd just move the `external` down to above the actual function, then the comment isn't necessary. The precise section containing `unlink` doesn't matter much.

dra27 commented on otherlibs/win32unix/unix.ml:
```diff
@@ -457,6 +458,21 @@ external chdir : string -> unit = "unix_chdir"
+  (* On Windows, trying to unlink a symlink to a directory will
+     raise, but the symlink can be deleted with rmdir. *)
+  try unlink path with
+  | Unix_error (EACCES, _, _) as exn_unlink ->
```
nit:
```suggestion
  | Unix_error (EACCES, _, _) as e ->
```
(cf. `realpath` above)

dra27 commented on otherlibs/win32unix/unix.ml:
```diff
@@ -457,6 +458,21 @@ external chdir : string -> unit = "unix_chdir"
+  try unlink path with
+  | Unix_error (EACCES, _, _) as exn_unlink ->
+    let {st_kind; _} = try lstat path with _ -> raise exn_unlink in
+    begin match st_kind with
+    | S_DIR ->
+      (* POSIX uses EPERM rather than EACCES when the path is a
+       * directory. *)
+      raise (Unix_error (EPERM, path, error_message EPERM))
+    | S_LNK -> rmdir path
+    | _ -> raise exn_unlink
+    end
```
```suggestion
    if st_kind = S_LNK then
      try rmdir path with _ -> raise e
    else
      raise e
```
(the key point is that `rmdir` will raise if the user actually doesn't have permission to delete the symlink, so you want to avoid leaking the `Unix.Unix_error(Unix.EACCES, "rmdir", "source")`)

MisterDA replied:
I thought that in that case it was more interesting to have the error that the user doesn't have the right permissions instead of the unlink error just telling us we're not using the right function to delete the entry.
I've seen that generally we want to re-raise the first error encountered, but here isn't the rmdir more relevant?

dra27 commented on runtime/sys.c:
```diff
@@ -286,10 +288,34 @@ CAMLprim value caml_sys_remove(value name)
   p = caml_stat_strdup_to_os(String_val(name));
   caml_enter_blocking_section();
   ret = unlink_os(p);
+#ifdef _WIN32
+  /* On Windows, trying to unlink a symlink to a directory will return
+   * EACCES, but the symlink can be deleted with rmdir. */
+  if (ret != 0 && errno == EACCES) {
+    if ((ret = GetFileAttributes(p)) != INVALID_FILE_ATTRIBUTES) {
+      if ((ret & FILE_ATTRIBUTE_REPARSE_POINT) &&
+          (ret & FILE_ATTRIBUTE_DIRECTORY)) {
+        ret = rmdir_os(p);
+      } else if (ret & FILE_ATTRIBUTE_DIRECTORY) {
+        /* POSIX uses EPERM rather than EACCES when the path is a
+         * directory. */
+        errno = EPERM;
+        ret = -1;
+      } else {
+        errno = EACCES;
+        ret = -1;
+      }
+    } else {
+      errno = EACCES;
+      ret = -1;
+    }
+  }
+#endif
```
This adds rather too much `_WIN32` to `sys.c`. `unlink_os` is sadly used in ocamlyacc, so I think the best course of action is to add a definition in `misc.h` for `caml_unlink` which will be `unlink` on Unix and `caml_win32_unlink` on Windows. You can then put the declaration of `caml_win32_unlink` in `osdeps.h` and move this function to `win32.c`.

If `otherlibs/unix/unlink.c` then uses `caml_unlink` instead of `unlink_os`, then the win32unix gets the change "for free" and I think all the changes in `otherlibs/win32unix/unix.ml` aren't needed.

dra27 commented on runtime/sys.c:
```diff
@@ -286,10 +288,34 @@ CAMLprim value caml_sys_remove(value name)
+      if ((ret & FILE_ATTRIBUTE_REPARSE_POINT) &&
+          (ret & FILE_ATTRIBUTE_DIRECTORY)) {
+        ret = rmdir_os(p);
+      } else if (ret & FILE_ATTRIBUTE_DIRECTORY) {
+        /* POSIX uses EPERM rather than EACCES when the path is a
+         * directory. */
+        errno = EPERM;
+        ret = -1;
+      } else {
+        errno = EACCES;
+        ret = -1;
+      }
+    } else {
+      errno = EACCES;
+      ret = -1;
+    }
```
In the separate PR, if you put the return value of `GetFileAttributes` into a different `DWORD`, then you'd have no need to update `ret` here. `errno` is only clobbered by `rmdir_os`, so there's never a need to set it to `EACCES`.

dra27 reviewed (commented):
There's a case for making the Windows `errno` either the same as Linux or at least the same as POSIX, but please do separate that, yes.

I think the overall premise of this PR is a good idea. The Windows behaviour is that `Unix.mkdir` does _more_ than on a POSIX system - i.e. it can delete a symlink to a directory, because that's how the OS works but we augment our portable `Sys.remove`/`Unix.unlink` to behave _at least_ as POSIX expects and remove symbolic links to directories as it would for files. Looking around us, libuv does similar (actually, libuv goes quite a lot further, but it its `unlink` function does remove directory symlinks!).

I think it will be simpler to do this entirely in C and implement `caml_win32_unlink` which can then be used by both the runtime and the Unix library.

MisterDA commented:
Thanks for the thorough review! I've implemented your suggestions, and removed the part where we raise `EPERM` when we try to delete a directory from this PR. I plan to squash the last two commits and keep the message of the last one ~when~ if the PR is accepted.

xavierleroy commented on Changes:
```diff
@@ -81,6 +85,10 @@ Working version
+- #10642: On Windows, Unix.unlink now removes symlinks to directories
+   instead of raising EACCES. (Antonin Décimo, review by David
+   Allsopp)
+
```
Why does this #10642 entry appears twice in Changes?  We don't do that.

MisterDA replied:
I did this because the PR changes `Sys.remove` from the runtime and `Unix.unlink` from win33unix which falls into the otherlibs section. If it's better not to duplicate, which section would be better suited for the entry?

dra27 replied:
There are quite a few instances of this where parts of related work fall into different sections of Changes.

xavierleroy commented on runtime/caml/misc.h:
```diff
@@ -356,6 +356,13 @@ extern double caml_log1p(double);
+#ifdef _WIN32
+#define caml_unlink caml_win32_unlink
+#else
+#define caml_unlink unlink_os
```
You could just `#define caml_unlink unlink` and remove the definition of `unlink_os` from `<caml/misc.h>`.

dra27 replied:
We still need the `unlink_os` definition for ocamlyacc, although it could be moved to ocamlyacc itself?

xavierleroy commented on runtime/win32.c:
```diff
@@ -797,6 +798,24 @@ int caml_win32_rename(const wchar_t * oldpath, const wchar_t * newpath)
+int caml_win32_unlink(const wchar_t * path) {
+  int ret, attrs;
+
+  ret = unlink_os(path);
```
Not sure why we're still using `*_os` macros in Win32-specific files.  What about calling `_wunlink` directly?

xavierleroy commented on runtime/win32.c:
```diff
@@ -797,6 +798,24 @@ int caml_win32_rename(const wchar_t * oldpath, const wchar_t * newpath)
+    if ((attrs = GetFileAttributes(path)) != INVALID_FILE_ATTRIBUTES &&
+        (attrs & FILE_ATTRIBUTE_REPARSE_POINT) &&
+        (attrs & FILE_ATTRIBUTE_DIRECTORY)) {
+      ret = rmdir_os(path);
```
Likewise, calling `_wrmdir` directly.

xavierleroy commented on runtime/win32.c:
```diff
@@ -797,6 +798,24 @@ int caml_win32_rename(const wchar_t * oldpath, const wchar_t * newpath)
+  int ret, attrs;
+
+  ret = unlink_os(path);
+  /* On Windows, trying to unlink a symlink to a directory will return
+   * EACCES, but the symlink can be deleted with rmdir. */
```
Sounds like a bug in Windows, but who am I to say.

dra27 replied:
It is certainly amusing that [MSDN](https://docs.microsoft.com/en-us/windows/win32/fileio/symbolic-links) claims "Microsoft has implemented its symbolic links to function just like UNIX links."! It's a consequence of Windows having two symbolic link types. IIRC you get `ERROR_ACCESS_DENIED` from `DeleteFile` for normal directories too.

xavierleroy reviewed (commented):
Looks reasonable to me, with some minor suggestions below.

@dra27: are you happy with this PR?

MisterDA commented:
> Sounds like a bug in Windows, but who am I to say.

My understanding is that on Windows, files and directories may have an associated data structure called a [reparse point](https://docs.microsoft.com/en-us/windows/win32/fileio/reparse-points). The reparse point is used to implement things such as "mounted folders" and symlinks. So a symlink-to-directory is really an empty directory with an associated reparse point, which explains why we need to use `rmdir` instead of `unlink`.
Reading more about that, I discovered the various uses of reparse points, and realized that this PR would allow to remove directories with all kinds of associated reparse point, which includes (but is not limited to) symlinks. It's possible to limit the PR to symlinks:

> use the FindFirstFile and FindNextFile functions to obtain the reparse tag in the dwReserved0 member of the WIN32_FIND_DATA structure. 

from [Determining Whether a Directory Is a Mounted Folder](https://docs.microsoft.com/en-us/windows/win32/fileio/determining-whether-a-directory-is-a-volume-mount-point), with the [`IO_REPARSE_TAG_SYMLINK`](https://docs.microsoft.com/en-us/windows/win32/fileio/reparse-point-tags) tag. I think we still want to be able to delete directories, regardless of the type of the associated reparse point... and maybe I should rephrase the Change entry to emphasize that?

MisterDA commented:
Oh, sorry, I got confused. We're talking about more-or-less emulating POSIX `unlink(2)` here, which forbids deletion of directories, but still allows deletion of symlinks-to-directories. So the conclusion of my previous comment is somewhat moot and I think I should definitely restrict this PR to target exclusively symlinks-to-directories and exclude directories with other kinds of reparse points.

dra27 commented:
Modulo the changes requested in in win32.c, this looks good to me, thanks! I'm not sure as to whether to move `unlink_os` specifically into ocamlyacc or not...

MisterDA commented:
With this new push I'm now using `FindFirstExW` to check whether the reparse point is a symlink, and I'm not using the `*_os` macros in `win32.c` anymore, however I've not removed `unlink_os` considering that ocamlyacc needs it.
Sorry about the two subsequent force-pushes, I got completely tricked by git!

dra27 commented:
> I'm now using `FindFirstExW` to check whether the reparse point is a symlink

This is a good catch - deleting junction points would be OK, but deleting mount points is not so good. However, `FindFirstFileExW` is tricky to get right - I'd suggest using `GetFileAttributes` as you were before but then have an extra check which opens the directory and uses `DeviceIoControl` as `otherlibs/win32unix/stat.c` does to be certain it's a symlink.

MisterDA commented:
Rebased the PR.

> You could just `#define caml_unlink unlink` and remove the definition of `unlink_os` from `<caml/misc.h>`.

I'm still not doing that because of ocamlyacc's dependency on `unlink_os`.

Introduce the `caml/winsupport.h` header to hold common code between
the runtime, lib-sys, and win32unix. This is also mandated by the name
conflict between `ATOM` from `<windef.h>` and `caml/instruct.h`.

I think the PR is ready, please let me know how I can further improve it if needed

dra27 commented:
I'm happy with this, thanks - @xavierleroy, I imagine "happy" won't be the word, but are you content for this to be merged? 🙂

xavierleroy commented:
Ill trust you (@dra27 and @MisterDA) on this Windows black magic.  I'd still like one Changes entry (in the "standard library" section) instead of two.

MisterDA commented:
> I'd still like one Changes entry (in the "standard library" section) instead of two.

I already had fixed that but not highlighted it in my messages, sorry.

dra27 commented on Changes:
```diff
@@ -100,6 +100,12 @@ Working version
+- #10642: On Windows, Sys.remove and Unix.unlink now remove symlinks
+  to directories instead of raising EACCES. Introduce
+  caml/winsupport.h to hold more common code between the runtime,
+  lib-sys, and win32unix. (Antonin Décimo, review by David Allsopp,
+  Xavier Leroy)
```
```suggestion
  lib-sys, and win32unix.
  (Antonin Décimo, review by David Allsopp and Xavier Leroy)
```

dra27 reviewed (commented):
Typo fix in Changes

dra27 committed:
Update Changes

xavierleroy commented:
> I already had fixed that but not highlighted it in my messages, sorry.

Thanks, and I'm the one who should apologize for not having checked the latest changes.

xavierleroy committed:
Merge branch 'trunk' into win32unix-posix-Sys.remove-Unix.unlink
